### PR TITLE
Add sizes to Index to use in Indexer removing Deprecated annotations

### PR DESCRIPTION
**Pull request by matteodg**
Getting back `sizes()` in `Indexer` and in `Index` as the maximum values of the indices.

### CODE REVIEW

File: src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java [hunk 1]
```diff
@@ -50,13 +50,19 @@
...
     public HyperslabIndex(long[] sizes, long[] offsets, long[] hyperslabStrides, long[] counts, long[] blocks) {
         super(sizes);
         this.offsets = offsets;
         this.hyperslabStrides = hyperslabStrides;
         this.counts = counts;
         this.blocks = blocks;
+
+        this.resultingSizes = new long[counts.length];
+        for (int i = 0; i < counts.length; i++) {
+            resultingSizes[i] = counts[i] * blocks[i];
```
[the remark this answers is not in the document]
Comment (matteodg) [reply src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java:64]: Just to not cause confusion with the `sizes` of `StrideIndex` as they are not the same as the resulting sizes of the hyperslab selection.

[the remark this answers is not in the document]
Comment (saudet) [reply src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java:64]: Ok, but there are not other `sizes` in `HyperslabIndex`, so I don't think they can be confused.

[the remark this answers is not in the document]
Comment (matteodg) [reply src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java:64]: The sizes in the `Index` are the ones used to reshape it, whereas the sizes in the `Hyperslab` are actually the result of a sub-selection from the underlying reshaped flat array/buffer/pointer.

Maybe the reshaping and sub-selection should actually be clearly separated.

[the remark this answers is not in the document]
Comment (saudet) [reply src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java:64]: Right, but the `sizes` that the user provides are used exactly no where in `HyperslabIndex`, so why do we care about them?

[the remark this answers is not in the document]
Comment (saudet) [reply src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java:64]: Oh, I see, they get used to compute the `strides`. Shouldn't users be allowed to set those then?

[the remark this answers is not in the document]
Comment (matteodg) [reply src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java]: Something like this? I added an `HyperslabIndex` constructor to allow to set the strides for reshaping.

[the remark this answers is not in the document]
Comment (saudet) [reply src/main/java/org/bytedeco/javacpp/indexer/HyperslabIndex.java]: Something like this yes, but let's make the tests not fail.

**Comment by saudet** [comment]
I've made a few changes such that `HyperslabIndex` actually just updates the `sizes`. I think that's OK since they do not get used by `Indexer`. In other words, it's only useful to users. What do you think?

**Comment by matteodg** [comment]
I just rebased on latest master and added generics to `reindex()`.

> I've made a few changes such that HyperslabIndex actually just updates the sizes. I think that's OK since they do not get used by Indexer. In other words, it's only useful to users. What do you think?

It is OK, I do not particularly like using `sizes` for two different purposes (reshaping and sub-selecting). And I guess you kept the `sizes` field and set to `index.sizes()` here:
```
    protected Indexer(Index index) {
        this.index = index;
        this.sizes = index.sizes();
        if (index instanceof StrideIndex) {
            this.strides = ((StrideIndex)index).strides();
        }
    }
```
so whoever extended `Indexer` can rely on the `sizes` protected field, that's fine with me at the end.

**Comment by saudet** [comment]
Yeah, it's not too clear what would be best here, but unless problems arise using the `sizes` field as output that way, I'd like to see how this goes. The thing is, we're not actually reshaping or selecting anything. When calling say `reindex()`, it's basically just a new view on the same data. We can think of it either as reshaping or (sub)selecting I suppose, but we're not actually differentiating between them. If we were to think of these 3 concepts separately as with NumPy, it's not always possible to use them together. For example, reshaping a slice doesn't always results in a view: Sometimes we need a copy. I'd rather just think of everything as views, at least for a simple API like Indexer. :)

BTW, we don't need the sizes in `Indexer` anymore because we can call `index.sizes()`. It's there just in case, but deprecated... Sounds OK?

**Comment by matteodg** [comment]
> If we were to think of these 3 concepts separately as with NumPy, it's not always possible to use them together. For example, reshaping a slice doesn't always results in a view: Sometimes we need a copy. I'd rather just think of everything as views, at least for a simple API like Indexer. :)

If I understand correctly with Foreign Memory API in JDK15 we will be able to create views in all cases (using a `VarHandle` to a Java array as well as to `Buffer`, `Pointer` or anything), that would be great!

One side question: are there somewhere the artifact of the current master branch? I cannot find them in Travis. I'm having troubles compiling for `linux-x86_64`: I tried to use the commands I found in `.travis.yml` but something is off (I still have to dig into it).

**Comment by saudet** [comment]
> If I understand correctly with Foreign Memory API in JDK15 we will be able to create views in all cases (using a `VarHandle` to a Java array as well as to `Buffer`, `Pointer` or anything), that would be great!

Not sure about whether or not it's going to be in JDK 15, but it will become possible at some point, yes. It still won't work with other implementations of Java such as Android though...
 
> One side question: are there somewhere the artifact of the current master branch? I cannot find them in Travis. I'm having troubles compiling for `linux-x86_64`: I tried to use the commands I found in `.travis.yml` but something is off (I still have to dig into it).

Yes, snapshot builds are available: http://bytedeco.org/builds/

**Comment by matteodg** [comment]
Thanks for the snapshot builds link, it works like a charm

**Pull request merged**